// Copyright (c) 2025 FRC 6328
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file at
// the root directory of this project.

package frc.robot.util;

import static edu.wpi.first.units.Units.*;

import java.util.ArrayList;
import java.util.List;

import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Pose3d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Rotation3d;
import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.units.measure.Distance;
import edu.wpi.first.wpilibj.DriverStation;
import frc.robot.Constants.MetaConstants;

public class AllianceFlipUtil {

  public static Distance applyX(Distance x) {
    return shouldFlip() 
        ? MetaConstants.Field.LENGTH.minus(x)  
        : x;
  }

  public static Distance applyY(Distance y) {
    return shouldFlip() 
        ? MetaConstants.Field.WIDTH.minus(y)
        : y;
  }

  public static Translation2d apply(Translation2d translation) {
    return new Translation2d(
      applyX(
        Meters.of(
          translation.getX()
        )
      ), 
      applyY(
        Meters.of(
          translation.getY()
        )
      )
    );
  }

  public static Rotation2d apply(Rotation2d rotation) {
    return shouldFlip() 
        ? rotation.rotateBy(Rotation2d.kPi) 
        : rotation;
  }

  // WARNING: i suck at rotation representation
  public static Rotation3d apply(Rotation3d rotation) {
    return shouldFlip() 
        ? new Rotation3d(
            -rotation.getX(),
            -rotation.getY(),
            -rotation.getZ()
        )
        : rotation;
  }

  public static Pose2d apply(Pose2d pose) {
    return shouldFlip()
        ? new Pose2d(apply(pose.getTranslation()), apply(pose.getRotation()))
        : pose;
  }

  public static ArrayList<Pose2d> apply(List<Pose2d> poses) {
    ArrayList<Pose2d> flippedPoses = new ArrayList<Pose2d>();

    for (Pose2d pose : poses) {
      flippedPoses.add(apply(pose));
    }

    return flippedPoses;
  }

  public static Pose3d apply(Pose3d pose) {
      return new Pose3d(
          applyX(
            Meters.of(
              pose.getX()
            )
          ),
          applyY(
            Meters.of(
              pose.getY()
            )
          ),
          Meters.of(pose.getZ()),
          apply(pose.getRotation())
      );
  }

  public static boolean shouldFlip() {
    return (DriverStation.getAlliance().isPresent()
        && DriverStation.getAlliance().get() == DriverStation.Alliance.Red);
  }
}
